### PR TITLE
Make Plugin easier to implement

### PR DESCRIPTION
**Pull request by EHER**
With this Refactory will be possible to Compose Xulieta with new file format Parsers or Code Validator.

It breaks BC

### CODE REVIEW

File: src/Parser/MarkdownParser.php [hunk 1]
```diff
@@ -2,34 +2,42 @@
...
 
-    private function __construct()
+    public function supportedExtensions(): array
     {
+        return ['markdown', 'md'];
+    }
+
+    public function isValid(SplFileInfo $file): bool
+    {
+        return in_array($file->getExtension(), $this->supportedExtensions(), false);
```
Comment (EHER) [inline src/Parser/MarkdownParser.php:33]: This should be `return true`.

A markdown file is valid even if the extension is not supported.

File: src/Parser/MultipleParser.php [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            }
+        }
+
+        return false;
+    }
+
+    public function isValid(SplFileInfo $file): bool
+    {
+        foreach ($this->parsers as $parser) {
+            if (! in_array($file->getExtension(), $parser->supportedExtensions())) {
```
Comment (malukenho) [inline src/Parser/MultipleParser.php]: It's missing the third argument.

File: src/Parser/MultipleParser.php [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+            }
+        }
+
+        return false;
+    }
+
+    public function getAllSampleCodes(SplFileInfo $file): array
+    {
+        foreach ($this->parsers as $parser) {
+            if (in_array($file->getExtension(), $parser->supportedExtensions())) {
```
Comment (malukenho) [inline src/Parser/MultipleParser.php:51]: It's missing the third argument.

File: src/Parser/Parser.php [hunk 4]
```diff
@@ -9,6 +9,12 @@
 
 interface Parser
 {
+    /** @psalm-return list<non-empty-string> */
+    public function supportedExtensions(): array;
+
+    public function supports(SplFileInfo $file): bool;
+
+    /** @TODO find a better name */
```
Comment (malukenho) [inline src/Parser/Parser.php]: Missing TODO

**Review by malukenho — CHANGES_REQUESTED** [review]

**Comment by malukenho** [comment]
Apart from DCO, it is looking good 👍🏻

**Review by malukenho — APPROVED** [review]

**Pull request merged**